File: rb/11.rb
nums = IO.read('../input/11').split.map(&:to_i)

max = 0

400.times do |i|
  [1, 19, 20, 21].each do |j|
    line = (0..j * 3).step(j)
    prod = line.map { |k| nums[i + k].to_i }.reduce(:*)
    max  = prod if prod > max
  end
end

p max

File: py/11.py
from re import split

with open('../input/11') as f:
    nums = map(int, split('\s', f.read())[:-1])

    max = 0

    for i in xrange(400):
        for j in [1, 19, 20, 21]:
            line = xrange(0, j * 4, j)
            prod = map(lambda k: nums[i + k if i + k < 400 else 0], line)
            prod = reduce(lambda a, b: a * b, prod)
            if prod > max:
                max = prod

print max
